core/server/services/auth/passwordreset.js

```diff
@@ -0,0 +1,153 @@
+const _ = require('lodash');
+const security = require('../../lib/security');
+const constants = require('../../lib/constants');
+const common = require('../../lib/common');
+const models = require('../../models');
+const urlUtils = require('../../lib/url-utils');
+const mail = require('../mail');
+
+const tokenSecurity = {};
+
+function generateToken(email, settingsAPI) {
+    const options = {context: {internal: true}};
+    let dbHash, token;
+
+    return settingsAPI.read(_.merge({key: 'db_hash'}, options))
+        .then((response) => {
+            dbHash = response.settings[0].value;
+
+            return models.User.getByEmail(email, options);
+        })
+        .then((user) => {
+            if (!user) {
+                throw new common.errors.NotFoundError({message: common.i18n.t('errors.api.users.userNotFound')});
+            }
+
+            token = security.tokens.resetToken.generateHash({
+                expires: Date.now() + constants.ONE_DAY_MS,
+                email: email,
+                dbHash: dbHash,
+                password: user.get('password')
+            });
+
+            return {
+                email: email,
+                resetToken: token
+            };
+        });
+}
+
+function extractTokenParts(options) {
+    options.data.passwordreset[0].token = security.url.decodeBase64(options.data.passwordreset[0].token);
```

Comment:
IMO would be nice to just have this function just accept the token - and have the consumer of the function deal with the mangling of the `options` object - maybe out of scope tho?

Reply:
Also - is there a reason that we mutate the object passed in?

Reply:
> and have the consumer of the function deal with the mangling of the options 

Was trying to keep the moves to the minimum, but bumped into this exact problem when reusing the function later in v2 migration. Maybe worth a small cleanup while on it after all. Will do it in this PR

> Also - is there a reason that we mutate the object passed in?

There's no good reason apart from the style of object mutations that were used throughout the system (just like how "frame" modifies the objects). Personally not a fan of mutations like this as they assume too much and it's impossible to compose such functions.

Reply:
> Maybe worth a small cleanup while on it after all.

:ok_hand: 

> Personally not a fan of mutations like this as they assume too much and it's impossible to compose such functions.

Yeah - IMO this is a pattern to avoid, it is prone to bugs, and usually hard to debug and promotes coupling between parts of a system. I don't think we need to do a whole refactor - but new code should probably not follow these practices - what do you think?

Reply:
> new code should probably not follow these practices

Completely agree. I'm debugging this thing right now and came out that validation function is mutating the options object to fit the data needed in these functions... it's a real PITA to decouple things without bigger code reshuffling: sweat_smile: 

Aiming to do a bare minimum of changes here but still getting a cleaner code. Will spend an hour on it and will merge whatever is there just to timeframe this somehow to avoid perfecting without a limit :smile:

Reply:
> it's a real PITA to decouple things without bigger code reshuffling

A _perfect_ example of why not to mutate arguments!

yeah man feel free to merge as is tbh - just worth raising so we're aware of it